wurst/objects/abilities/mage/unsub/Wabbajack.wurst

```diff
@@ -0,0 +1,168 @@
+        ..setBuffs(1, commaList(BUFF_ID, BUFF_ID_AIR))
+
+class Wabbajack extends ChannelAbilityPreset
+    construct(int newAbilityId, string hotkey)
+        super(newAbilityId, 1, true)
+        this.setIconNormal(Icons.bTNWitchDoctorMaster)
+        this.setIconResearch(Icons.bTNWitchDoctorMaster)
+        this.setIconTurnOff(Icons.bTNWitchDoctorMaster)
+        this.setHeroAbility(false)
+        this.setHotkeyNormal(hotkey)
+        this.setName("Wabbajack's Tormenter")
+        this.setDisableOtherAbilities(1, false)
+        this.setCastingTime(1, 0)
+        this.setTooltipNormal(1,"Wabbajack's Tormenter".color(GENERAL_COLOR))
+        this.setTooltipNormalExtended(1, TOOLTIP_EXT)
+        this.setCooldown(1, COOLDOWN)
+        this.setManaCost(1, MANACOST)
+        this.presetTargetTypes(Targettype.POINT)
+        this.setButtonPositionNormalX(1)
+        this.setButtonPositionNormalY(1)
+        this.setCastRange(1, MISSILE_SPEED * MISSILE_LIFETIME)
+        this.setFollowThroughTime(1, 0.5)
+
+@compiletime function createWabbajack()
+    new Wabbajack(ABILITY_WABBAJACK, "Q")
+
+class WabbajackProjectile
+    use TimedLoop
+    real lifeTime
+    unit ownerUnit
+    angle ang
+    vec3 pos
+    effect efx
+    bool exploded
+    bool isOvercharged
+    unit target
+
+    construct(unit caster, vec2 casterPos, angle ang)
+        this.ownerUnit = caster
+        this.lifeTime = 0
+        this.ang = ang
+        this.pos = casterPos.polarOffset(ang, SPAWN_OFFSET_DISTANCE).withTerrainZ(MISSILE_FLY_HEIGHT)
+        this.efx = addEffect(Abilities.zigguratMissile, pos)
+        this.exploded = false
+        this.efx.setRoll(ang)
+        this.target = null
+        this.startTimedLoop()
+
+    override function onTimedLoop()
+        if this.exploded or this.lifeTime >= MISSILE_LIFETIME
+            stopTimedLoopAndDestroy()
+        this.lifeTime += LOOP_INTERVAL
+        vec3 newPos = this.pos.offset2d(this.ang, MISSILE_SPEED * LOOP_INTERVAL)
+        let terrainZ = newPos.getTerrainZ()
+        newPos.z = terrainZ + MISSILE_FLY_HEIGHT
+        this.efx.setPos(newPos)
+        this.pos = newPos
+        checkForImpacts()
+
+    function checkForImpacts()
+        forUnitsInRange(this.pos.toVec2(), IMPACT_TRIGGER_AOE) u ->
+            if isValidTarget(ownerUnit, u)
+                this.target = u
+                if u.isTroll()
+                    spawnTemporaryVision(VISION_RANGE, DURATION_HERO)
+                else
+                    spawnTemporaryVision(VISION_RANGE, DURATION_NORMAL)
+                onImpact()
+
+    function onImpact()
+        this.exploded = true
+        InstantDummyCaster.castTarget(this.ownerUnit.getOwner(), ABILITY_WABBAJACK_CAGE, 1, Orders.ensnare, this.target)
+
+    function spawnTemporaryVision(real radius, real duration)
+        let vis = createVision(ownerUnit.getOwner(), this.pos.toVec2(), radius, true)
+        vis.start()
+        doAfter(duration) ->
+            vis.destr()
```

Comment:
This feels weird if the unit is released early, either by dying or being cured (assuming that removes the effect). You could use a fake faerie ability without armor reduction or an art effect?

Reply:
True.

Reply:
However there is no field to set a vision range with faerie fire, it uses the target vision range, I don't really want the vision range to be too big and give for example, vision of the target base. I just want the spell caster to know that he hit a target, the effect last 1.75 sec on hero so I think that's negligeable.

Reply:
OK.